**### ---------  TASK-1 : LOAD & INSPECT ----------  ###**

In [149]:
# Read the csv file

with open('bangalore_tech_salaries.csv','r', encoding='utf-8') as f :
  content = f.read()


In [150]:
# read file line by line

with open('bangalore_tech_salaries.csv', 'r', encoding='utf-8') as f:
   lines = f.readlines()

In [151]:
import pandas as pd

df = pd.read_csv('bangalore_tech_salaries.csv')
df

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite
...,...,...,...,...,...,...,...,...,...,...,...,...
1010,BLR0824,Product Designer,2,17.6,12.4,Zoho,Mid-size,"React, SQL, ML, Deep Learning, Java",Jayanagar,Tier-2,2022,WFH
1011,BLR0792,Product Lead,4,₹40.2 LPA,26.6,HCL,MNC,"Excel, Java, Deep Learning, Python, JavaScript",Whitefield,Tier 2,2021,Work from Office
1012,BLR0209,BI Analyst,1,8.6 LPA,6.0 LPA,EdMint,Early-stage,"PyTorch, System Design, MongoDB, JIRA, PowerBI...",NaN,2,2024,Hybrid
1013,BLR0788,PM,2,27.6,17.4,Zomato,Unicorn,"Azure, Kafka, Excel",Bellandur,Tier-3,2023,Remote


In [154]:
# Let's inspect the file
# it shows info about dataset like columns, their names, non-null values, data types, etc.

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB


In [155]:
# identify the shape of dataset
df.shape

# it means there are total rows=1015 and columns=12 in the entire dataset

(1015, 12)

In [156]:
# let's see the different data types present in the dataset

df.dtypes

# it means
# years_exp and joining_year this are int64 only
# other columns include objects

,0
Employee_ID,object
Role,object
years_exp,int64
Current_CTC,object
Previous_CTC,object
Company,object
company_TYPE,object
Skills,object
Location,object
Education_Tier,object


In [157]:
# let's count total columns
df.columns

column_count = df.columns.value_counts().sum()
print(column_count)

# there are total 12 columns in the dataset

12


In [158]:
# let's see how many null values are there in the dataset

df.isnull()

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,True,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...
1010,False,False,False,False,False,False,False,False,False,False,False,False
1011,False,False,False,False,False,False,False,False,False,False,False,False
1012,False,False,False,False,False,False,False,False,True,False,False,False
1013,False,False,False,False,False,False,False,False,False,False,False,False


In [159]:
# let's see total count of null values
df.isnull().sum()

# So there are 200 null values in Previous_CTC column and
# 27 null values in Skills column
# and 20 null values in Location column

,0
Employee_ID,0
Role,0
years_exp,0
Current_CTC,0
Previous_CTC,200
Company,0
company_TYPE,0
Skills,27
Location,20
Education_Tier,0


In [160]:
# columns with most missing data

column_missing_values = df.isnull().sum().sort_values(ascending=False)
print(column_missing_values[0])

# find columns name which has most missing values

column_name_miss = column_missing_values.index[0]
print(column_name_miss)

200
Previous_CTC


/tmp/ipykernel_1225/2433783030.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(column_missing_values[0])


In [161]:
# let's see duplicate values now
df.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
1010,False
1011,False
1012,False
1013,False


In [162]:
# count total duplicate values

df.duplicated().sum()

# So there are total 15 duplicate values in entire dataset

np.int64(15)

In [163]:
# Let's print a short summary of what we have find till now

print("==============================================================")
print("                      OVERALL OVERVIEW                          ")
print("==============================================================")

print()
print(f"Total no.of columns     = {column_count}")
print(f"Total no.of rows        = {df.shape[0]}")
print(f"Total no.of duplicates  = {df.duplicated().sum()}")
print(f"Total no.of null values = {df.isnull().sum().sum()}")
print(f"Column with most missing values = {column_name_miss} : {column_missing_values[0]}")

print()
print("==============================================================")
print("==============================================================")

                      OVERALL OVERVIEW                          

Total no.of columns     = 12
Total no.of rows        = 1015
Total no.of duplicates  = 15
Total no.of null values = 247
Column with most missing values = Previous_CTC : 200



/tmp/ipykernel_1225/357277283.py:12: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Column with most missing values = {column_name_miss} : {column_missing_values[0]}")


**### ----------------  TASK-2 : CLEAN ------------------  ###**

In [164]:
# let's again see through the dataset

import pandas as pd
import numpy as np

df = pd.read_csv('bangalore_tech_salaries.csv')
df.head()

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite


In [165]:
# let's rename some column names
# rename columns to snake_case

df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
)
df.columns       # so we made column names into lowercase and analysis suitable format
                # example :  Current CTC -> current_ctc   like this

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')

In [72]:
# standardize the role names now

df['role'].head(10)

,role
0,DS
1,DevOps
2,SDE Backend
3,SDE Frontend
4,Product Lead
5,Product Manager
6,Site Reliability Engineer
7,DA
8,DS
9,UI/UX


In [166]:
df['role'].tail(10)

,role
1005,Sr PM
1006,SRE
1007,Backend Developer
1008,SDE Frontend
1009,Designer
1010,Product Designer
1011,Product Lead
1012,BI Analyst
1013,PM
1014,DA


In [167]:
role_mapping = {
    "DA" : "Data Analyst",
    "BE" : "Backend Engineer",
    "BA" : "BI Analyst",
    "SRE" : "Site Reliability Engineer",
    "FE" : "Frontend Engineer",
    "DS" : "Data Scientist",
    "Sr PM" : "Senior Product Manager",
    "PM" : "Product Manager",
    "SDE FS" : "SDE Full Stack"
}

df["role_clean"] = df["role"].replace(role_mapping)
df["role_clean"]

df["role_clean"].value_counts()

,count
role_clean,
BI Analyst,71
Data Scientist,66
Site Reliability Engineer,61
Product Manager,55
Data Analyst,51
Business Analyst,37
Backend Engineer,36
Business Systems Analyst,32
Product Lead,32


In [168]:
# standardize the company type now

df['company_type'].head(10)

,company_type
0,Unicorn
1,MNC
2,Mid-size
3,MNC
4,early-stage
5,early-stage
6,MNC
7,MNC
8,Mid-size
9,Mid-size


In [169]:
df['company_type'].tail(10)

,company_type
1005,Mid-size
1006,Unicorn
1007,Unicorn
1008,early-stage
1009,unicorn
1010,Mid-size
1011,MNC
1012,Early-stage
1013,Unicorn
1014,MNC


In [170]:
# standardize company name

company_mapping = {
    "early-stage" : "Early-stage",
    "EARLY-STAGE" : "Early-stage",
    "mid-size" : "Mid-size",
    "MID-SIZE" : "Mid-size",
    "Mid_size" : "Mid-size",
    "unicorn" : "Unicorn",
    "UNICORN" : "Unicorn",
    "mnc" : "MNC"
}

df["company_type_clean"] = df["company_type"].replace(company_mapping)
df["company_type_clean"]

df["company_type_clean"].value_counts()

,count
company_type_clean,
MNC,348
Mid-size,261
Unicorn,207
Early-stage,199


In [171]:
# standardize the education tier now

df['education_tier'].head(10)

,education_tier
0,Tier-1
1,2
2,Tier-1
3,Tier-2
4,Tier-2
5,Tier-2
6,Tier 1
7,Tier-2
8,Tier 2
9,T3


In [172]:
# standardize education tier name

education_mapping = {
    "2" : "Tier-2",
    "T2" : "Tier-2",
    "Tier 2" : "Tier-2",
    "1" : "Tier-1",
    "T1" : "Tier-1",
    "Tier 1" : "Tier-1",
    "3" : "Tier-3",
    "T3" : "Tier-3",
    "Tier 3" : "Tier-3"
}

df["education_tier_clean"] = df["education_tier"].replace(education_mapping)
df["education_tier_clean"]

df["education_tier_clean"].value_counts()

,count
education_tier_clean,
Tier-2,510
Tier-3,309
Tier-1,196


In [173]:
# convert current CTC strings into Float (LPA)

df['current_ctc'].head(10)

,current_ctc
0,49.4
1,9.7
2,"3,360,000"
3,41.4 LPA
4,"3,640,000"
5,32.1
6,24.6
7,21.7 LPA
8,21.0
9,26.9 LPA


In [174]:
# convert CTC strings into Float (LPA)


df['current_ctc'] = (df['current_ctc'].str.strip()
.str.replace('LPA', '', regex = False)
.str.replace('lpa', '', regex = False)
.str.replace('Lpa', '', regex = False)
.str.replace('₹', '', regex = False)
.str.replace(',', '', regex = False)
.str.replace(' ', '', regex = False)
)


df['current_ctc'].head(10)

,current_ctc
0,49.4
1,9.7
2,3360000
3,41.4
4,3640000
5,32.1
6,24.6
7,21.7
8,21.0
9,26.9


In [175]:
# convert previous CTC strings into Float (LPA)


df['previous_ctc'] = (df['previous_ctc'].str.strip()
.str.replace('LPA', '', regex = False)
.str.replace('lpa', '', regex = False)
.str.replace('Lpa', '', regex = False)
.str.replace('₹', '', regex = False)
.str.replace(',', '', regex = False)
.str.replace(' ', '', regex = False)
)


df['previous_ctc'].head(10)

,previous_ctc
0,32.4
1,NaN
2,24.2
3,30.3
4,30.6
5,22.0
6,19.3
7,17.7
8,NaN
9,18.9


In [176]:
# convert columns to numeric

df['current_ctc'] = pd.to_numeric(df['current_ctc'])
df['previous_ctc'] = pd.to_numeric(df['previous_ctc'])

# drop duplicates
df = df.drop_duplicates()
print(df.shape)

(1000, 15)


In [177]:
# handle missing values
df['current_ctc'] = df['current_ctc'].fillna(df['current_ctc'].mean())
df['previous_ctc'] = df['previous_ctc'].fillna(df['previous_ctc'].mean())

df['role'] = df['role_clean'].fillna(df['role_clean'].mode()[0])
df['company_type'] = df['company_type_clean'].fillna(df['company_type_clean'].mode()[0])
df['education_tier'] = df['education_tier_clean'].fillna(df['education_tier_clean'].mode()[0])

df

print(df.dtypes)

employee_id              object
role                     object
years_exp                 int64
current_ctc             float64
previous_ctc            float64
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
company_type_clean       object
education_tier_clean     object
dtype: object


In [178]:
# final output

print("\nData Types")
print(df.dtypes)

print("\nRole Counts")
print(df['role_clean'].value_counts())

print("\nCompany Type counts")
print(df['company_type_clean'].value_counts())

print("\nEducation Tier counts")
print(df['education_tier_clean'].value_counts())




Data Types
employee_id              object
role                     object
years_exp                 int64
current_ctc             float64
previous_ctc            float64
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
company_type_clean       object
education_tier_clean     object
dtype: object

Role Counts
role_clean
BI Analyst                   69
Data Scientist               64
Site Reliability Engineer    61
Product Manager              55
Data Analyst                 50
Backend Engineer             36
Business Analyst             35
Business Systems Analyst     32
Frontend Engineer            31
Product Lead                 31
ML Engineer                  31
Data Science Engineer        30
Designer                     28
Senior Product Manager       26
Frontend Developer    

**### ----------------  TASK-3 : FIVE BUSINESS QUESTIONS ------------------  ###**

In [179]:
''' Q3.1 — CTC distribution by role. For every role, compute median, mean, min, and max CTC. Sort roles by
median descending. Which role pays the most? Which the least? Is the mean materially different from
the median in any role (sign of outliers)?

Approach :

I grouped the dataset by cleaned job role and calculated the mean, median, minimum and maximum current CTC for each role.
Then I sorted thre results by median salary in descending order to compare salary distributions across roles.

'''

' Q3.1 — CTC distribution by role. For every role, compute median, mean, min, and max CTC. Sort roles by\nmedian descending. Which role pays the most? Which the least? Is the mean materially different from\nthe median in any role (sign of outliers)?\n\nApproach :\n\nI grouped the dataset by cleaned job role and calculated the mean, median, minimum and maximum current CTC for each role.\nThen I sorted thre results by median salary in descending order to compare salary distributions across roles.\n\n'

In [180]:
ctc_stats = df.groupby("role_clean")["current_ctc"].agg(["mean", "median", "min", "max"])
ctc_stats = ctc_stats.sort_values(by="median", ascending=False)
print(ctc_stats)

# AI-assisted: Used ChatGPT to help write the code for salary analysis.

                                    mean  median   min        max
role_clean                                                       
Senior Product Manager     367339.419231   40.05  10.8  4730000.0
Product Lead               662927.316129   34.10  12.0  4370000.0
Product Manager            615665.625455   32.90  13.8  5690000.0
ML Engineer                174867.296774   31.40  10.0  2220000.0
Data Scientist             496270.885937   29.15  10.9  4170000.0
Data Science Engineer      364358.046667   28.15  14.3  2900000.0
Frontend Developer         341270.191667   26.95  11.0  2930000.0
UI/UX                      147411.395652   26.90   6.2  1590000.0
Backend Engineer           366689.175000   26.35   9.9  4040000.0
SDE Frontend               243518.570000   26.15   6.7  3060000.0
Site Reliability Engineer  405101.683607   25.50   8.8  4840000.0
SDE Full-Stack             247526.093750   25.45  12.5  2890000.0
SDE Full Stack             710854.837500   25.25  11.5  7090000.0
Infra Engi

In [181]:
# let's see which role pays the most

print("Role with highest median salary : ")
print(ctc_stats.head(1))

Role with highest median salary : 
                                 mean  median   min        max
role_clean                                                    
Senior Product Manager  367339.419231   40.05  10.8  4730000.0


In [120]:
# let's see which role pays the least

print("Role with lowest median salary : ")
print(ctc_stats.tail(1))

Role with lowest median salary : 
                   mean  median  min        max
role_clean                                     
Data Analyst  186015.27    16.9  5.6  2790000.0


In [182]:
# check the outliers

ctc_stats["difference"] = ctc_stats["mean"] - ctc_stats["median"]
print(ctc_stats.sort_values(by="difference",ascending=False))

                                    mean  median   min        max  \
role_clean                                                          
SDE Full Stack             710854.837500   25.25  11.5  7090000.0   
Product Lead               662927.316129   34.10  12.0  4370000.0   
Product Manager            615665.625455   32.90  13.8  5690000.0   
BI Analyst                 588130.375362   19.00   5.2  5270000.0   
Data Scientist             496270.885937   29.15  10.9  4170000.0   
DevOps Engineer            432521.087500   21.80   9.8  5610000.0   
Site Reliability Engineer  405101.683607   25.50   8.8  4840000.0   
Designer                   387158.189286   18.85   7.1  4380000.0   
Senior Product Manager     367339.419231   40.05  10.8  4730000.0   
Backend Engineer           366689.175000   26.35   9.9  4040000.0   
Data Science Engineer      364358.046667   28.15  14.3  2900000.0   
UI Designer                357516.350000   20.55   8.6  2940000.0   
Full Stack Engineer        357294.

**Insight :

1.   SDE Full Stack         -    710829.587500
2.   Product Lead            -   662893.216129
3.   Product Manager          -  615632.725455
4.   shows the largest gap between mean and median CTC, suggesting the presence of high-salary outliers.
  **




In [ ]:
''' Q3.2 — Experience curve for SDE Backend. Bin SDE Backend professionals by years_exp into 4 buckets
(0-1, 2-3, 4-5, 6+). Compute median CTC per bucket. What is the CTC growth per experience band? Hint:
pd.cut + groupby.

Approach :

I filtered the dataset for SDE Backend professionals, grouped them into four experience buckets using pd.cut()
and calculated the median current CTC for each bucket to analyze salary growth with experience.

In [184]:
# Filter only SDE Backend professionals
backend_df = df[df["role_clean"] == "SDE Backend"].copy()

# Create experience buckets
bins = [0, 2, 4, 6, float("inf")]
labels = ["0-1", "2-3", "4-5", "6+"]

backend_df["experience_band"] = pd.cut(
    backend_df["years_exp"],
    bins=bins,
    labels=labels,
    right=False
)

# Calculate median CTC for each experience band
median_ctc = (
    backend_df.groupby("experience_band")["current_ctc"]
    .median()
)

print(median_ctc)

experience_band
0-1    13.95
2-3    23.30
4-5    27.60
6+     46.00
Name: current_ctc, dtype: float64


/tmp/ipykernel_1225/3702323632.py:17: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  backend_df.groupby("experience_band")["current_ctc"]


**Insight:**

1.   The median CTC for SDE Backend professionals increases consistently with experience, from 13.95 LPA (0–1 years) to 46.00 LPA (6+ years).
2.   The highest salary growth is observed for professionals with more than 6 years of experience, indicating that experience has a strong positive impact on compensation.


In [ ]:
''' Q3.3 — Skill premium for SDEs. Pick 4 skills from {AWS, ML, System Design, Kubernetes, Deep Learning,
Microservices}. For each, compare median CTC of SDEs (any of Backend / Frontend / Full-Stack) who
have the skill vs those who don't. Which skill adds the most to median CTC? Hint: .str.contains on the
cleaned skills column

Approach :

I filtered the dataset to include only SDE professionals (Backend, Frontend, and Full-Stack).
Then, for each selected skill, I divided professionals into two groups—those who have the skill and those who do not.
Finally, I calculated the median current CTC for both groups
and compared the salary difference to identify the skill with the highest salary premium. '''

In [185]:
df.columns

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode', 'role_clean', 'company_type_clean',
       'education_tier_clean'],
      dtype='object')

In [186]:
# Filter only SDE professionals
sde_df = df[
    df["role_clean"].str.contains(
        "Backend|Frontend|Full.?Stack",
        case=False,
        na=False
    )
].copy()

skills = ["AWS", "ML", "System Design", "Deep Learning"]

results = []

for skill in skills:

    has_skill = sde_df[
        sde_df["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc"].median()

    no_skill = sde_df[
        ~sde_df["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc"].median()

    premium = has_skill - no_skill

    results.append([skill, has_skill, no_skill, premium])

result_df = pd.DataFrame(
    results,
    columns=[
        "Skill",
        "Median CTC (Has Skill)",
        "Median CTC (No Skill)",
        "Premium"
    ]
)

result_df = result_df.sort_values(by="Premium", ascending=False)

print(result_df)

           Skill  Median CTC (Has Skill)  Median CTC (No Skill)  Premium
2  System Design                   26.10                   23.3     2.80
1             ML                   25.45                   23.6     1.85
0            AWS                   25.10                   23.5     1.60
3  Deep Learning                   23.60                   24.0    -0.40


In [187]:
print("\nSkill with Highest Premium:")
print(result_df.iloc[0])


Skill with Highest Premium:
Skill                     System Design
Median CTC (Has Skill)             26.1
Median CTC (No Skill)              23.3
Premium                             2.8
Name: 2, dtype: object


**Insight:**



1.   Among the selected skills, System Design provides the highest median CTC premium of 2.80 LPA for SDE professionals, followed by ML (1.85 LPA) and AWS (1.60 LPA).
2.   Deep Learning shows a slightly negative premium (-0.40 LPA), indicating no salary advantage for this dataset.



In [ ]:
''' Q3.4 — Company-type premium, same role. Pick one role (SDE Backend or Data Scientist). Compute
median CTC across Unicorn / MNC / Mid-size / Early-stage for that role. What is the percentage
premium of Unicorn over MNC at the same role?

Approach :

I selected the Data Scientist role and grouped the data by company type.
Then I calculated the median current CTC for each company type
and computed the percentage salary premium of Unicorn companies over MNCs.
'''

In [188]:
# Filter Data Scientist role
role_df = df[df["role_clean"] == "Data Scientist"]

# Median CTC by company type
median_ctc = (
    role_df.groupby("company_type_clean")["current_ctc"]
    .median()
)

print(median_ctc)

company_type_clean
Early-stage    30.80
MNC            24.70
Mid-size       23.90
Unicorn        46.05
Name: current_ctc, dtype: float64


In [189]:
# calculate unicorn premium over MNC

unicorn = median_ctc["Unicorn"]
mnc = median_ctc["MNC"]

premium = ((unicorn - mnc) / mnc) * 100

print(f"Unicorn Premium over MNC: {premium:.2f}%")

Unicorn Premium over MNC: 86.44%


**Insight:**



1.   Data Scientists working in Unicorn companies have the highest median CTC.
2.   Their median salary is 86.44% higher than that of Data Scientists working in MNCs, showing a clear salary premium for Unicorn companies.


In [ ]:
''' Q3.5 — Underpaid professionals (multi-step). For every (role_clean, company_type_clean, exp_band)
combination, compute median CTC. Then for every individual, compute the gap = their_ctc -
their_group_median. Print the top 10 most-underpaid professionals with name, role, company_type,
exp, ctc, group_median, gap. Filter to groups with at least 10 members (so the median is reliable). Hint:
groupby + transform + filter.

Approach :

I grouped employees based on role, company type, and experience band.
I calculated the median CTC for each group and assigned it to every employee.
Then I computed the salary gap (Current CTC − Group Median), filtered groups with at least 10 employees
and displayed the top 10 employees with the largest negative salary gap.

'''

In [190]:
# Create experience bands

bins = [0, 2, 4, 6, float("inf")]
labels = ["0-1", "2-3", "4-5", "6+"]

df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=bins,
    labels=labels,
    right=False
)

In [191]:
# compute group median

group_cols = ["role_clean", "company_type_clean", "exp_band"]

df["group_median"] = (
    df.groupby(group_cols)["current_ctc"]
      .transform("median")
)

/tmp/ipykernel_1225/2175689789.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(group_cols)["current_ctc"]


In [140]:
# count members in each group

df["group_size"] = (
    df.groupby(group_cols)["employee_id"]
      .transform("count")
)

# compute salary gap
df["gap"] = df["current_ctc"] - df["group_median"]

# filter reliable groups
filtered_df = df[df["group_size"] >= 10]

# AI-assisted: Used ChatGPT to understand grouping and aggregation.

/tmp/ipykernel_1225/111373473.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(group_cols)["employee_id"]


In [141]:
# Top 10 underpaid employees

underpaid = (
    filtered_df.sort_values("gap")
    .head(10)
)

print(
    underpaid[
        [
            "employee_id",
            "role_clean",
            "company_type_clean",
            "exp_band",
            "current_ctc",
            "group_median",
            "gap"
        ]
    ]
)

    employee_id      role_clean company_type_clean exp_band  current_ctc  \
761     BLR0887  Data Scientist                MNC      2-3         19.7   
203     BLR0907  Data Scientist                MNC      2-3         19.8   
248     BLR0757  Data Scientist                MNC      2-3         20.7   
135     BLR0120  Data Scientist                MNC      2-3         21.8   
267     BLR0207      BI Analyst                MNC      2-3         12.3   
521     BLR0503      BI Analyst                MNC      2-3         13.8   
751     BLR0600      BI Analyst                MNC      2-3         14.0   
787     BLR0200      BI Analyst                MNC      2-3         14.4   
592     BLR0921  Data Scientist                MNC      2-3         24.2   
583     BLR0073      BI Analyst                MNC      2-3         16.7   

     group_median  gap  
761          26.6 -6.9  
203          26.6 -6.8  
248          26.6 -5.9  
135          26.6 -4.8  
267          16.9 -4.6  
521          

**Insight:**

1.   The analysis identifies employees whose current CTC is below the median salary of professionals with the same role, company type, and experience band.
2.   The most underpaid employee is BLR0887 (Data Scientist, MNC, 2–3 years), earning 19.7 LPA, which is 6.9 LPA below the group median of 26.6 LPA.





**### ----------------  TASK-4 : BUILD THE PRINTED REPORT ------------------  ###**




In [148]:
# ===========================
# AI-assisted: Generated formatted report header using Python print formatting.
# ===========================

print("=" * 70)
print(" " * 18 + "BANGALORE TECH SALARY DECODER")
print("=" * 70)

print(f"{'Built by':<12}: Upasna Sunil Bhostekar")
print(f"{'Project':<12}: Unlock Academy 2-Hour Live Project")
print(f"{'Dataset':<12}: Engineering Professionals (Synthetic)")
print(f"{'Period':<12}: 2024 Employment Snapshot")
print()
print("=" * 70)


# median by CTC role
print("-" * 70)
print("MEDIAN CTC BY ROLE (LPA)")
print("-" * 70)

# Sort by median salary
role_report = ctc_stats.sort_values(by="median", ascending=False)

# Print top 8 roles with ASCII bars
for role, row in role_report.head(8).iterrows():

    median_salary = row["median"]

    # Create ASCII bar (1 block ≈ 2 LPA)
    bar = "█" * int(median_salary / 2)

    print(f"{role:<25} {bar:<20} {median_salary:>6.2f}")

print()


# SDE backend experience brand
print("-" * 70)
print("SDE BACKEND CTC BY EXPERIENCE BAND")
print("-" * 70)

for band, ctc in median_ctc.items():
    print(f"{band:<10} : {ctc:>6.2f} LPA")

print()


# skill premium for SDES
print("-" * 70)
print("SKILL PREMIUM FOR SDEs")
print("-" * 70)

for _, row in result_df.iterrows():
    print(f"{row['Skill']:<20} +{row['Premium']:>5.2f} LPA")

print()
print(f"Highest Skill Premium : {result_df.iloc[0]['Skill']} ({result_df.iloc[0]['Premium']:.2f} LPA)")
print()


# company type premium
print("-" * 70)
print("COMPANY TYPE PREMIUM (DATA SCIENTIST)")
print("-" * 70)

for company, ctc in median_ctc.items():
    print(f"{company:<15} : {ctc:>6.2f} LPA")

print()
print(f"Unicorn Premium over MNC : {premium:.2f}%")
print()


# Top unpaid preofessionals
print("-" * 70)
print("TOP UNDERPAID PROFESSIONAL")
print("-" * 70)

top = underpaid.iloc[0]

print(f"Employee ID : {top['employee_id']}")
print(f"Role        : {top['role_clean']}")
print(f"Company     : {top['company_type_clean']}")
print(f"Experience  : {top['exp_band']}")
print(f"Current CTC : {top['current_ctc']:.2f} LPA")
print(f"Group Median: {top['group_median']:.2f} LPA")
print(f"Gap         : {top['gap']:.2f} LPA")
print()


# let's end the report with footer
print("=" * 70)
print(" " * 27 + "END OF REPORT")
print("=" * 70)

# AI-assisted: Used ChatGPT to improve report formatting using f-strings.

                  BANGALORE TECH SALARY DECODER
Built by    : Upasna Sunil Bhostekar
Project     : Unlock Academy 2-Hour Live Project
Dataset     : Engineering Professionals (Synthetic)
Period      : 2024 Employment Snapshot

----------------------------------------------------------------------
MEDIAN CTC BY ROLE (LPA)
----------------------------------------------------------------------
Senior Product Manager    ████████████████████  40.05
Product Lead              █████████████████     34.10
Product Manager           ████████████████      32.90
ML Engineer               ███████████████       31.40
Data Scientist            ██████████████        29.15
Data Science Engineer     ██████████████        28.15
Frontend Developer        █████████████         26.95
UI/UX                     █████████████         26.90

----------------------------------------------------------------------
SDE BACKEND CTC BY EXPERIENCE BAND
--------------------------------------------------------------------

**### ----------------  TASK-5 : WRITE THREE INSIGHTS - IN YOUR OWN WORDS ------------------  ###**




### Insight 1
1. System Design provides the highest salary premium among the selected SDE skills, increasing the median CTC by **2.80 LPA** compared to professionals without this skill.
2. Students aiming for high-paying SDE roles should focus on mastering System Design.

### Insight 2
1. For SDE Backend professionals, the median CTC increases from **13.95 LPA** (0–1 years) to **46.00 LPA** (6+ years).
2. This shows that gaining experience improves salary, so students should focus on long-term skill development and career growth.

### Insight 3
1. Data Scientists working in Unicorn companies earn **86.44%** higher median CTC than those in MNCs for the same role.
2. Students should consider applying to Unicorn companies after building strong technical skills and experience.